File: 001_array_linked_list.py


##########  ##########  ##########  ##########  ##########
class ListNode() :
    def __init__(self, val) :
        self.val = val
        self.nxt = None

def list_to_num(lst) :
    digit = 1
    num = 0
    while lst :
        num += lst.val*digit
        digit *= 10
        lst = lst.nxt
    print(num)

def num_to_list(num) :
    dummy=pivot=ListNode(-1)
    while num :
        pivot.nxt = ListNode(num%10)
        num //= 10
        pivot = pivot.nxt
    return dummy.nxt

def addNode(list, val) :
    # new node with data
    # new node point to what
    # what point to new node
    if list is None : return list
    dummy=tail=ListNode(-1)
    dummy.nxt=list
    while tail.nxt :
        tail=tail.nxt
    tail.nxt=ListNode(val)
    return dummy.nxt


def removeNode(list) :
    # find the node by node.nxt
    # modify the last node's next point

########## Easy ##########  ##########  ##########  ##########

class lc001 : #E, 2sum
    def n2 (self, nums, target) :
        for i, num_i in enumerate(nums) :
            for j, num_j in enumerate(nums[i:]) :
                if num_i+num_j == target :
                    return [i, j]
        return [-1, -1]
    
    def n (self, nums, target) :
        dic = {}
        for i, num in enumerate(nums) :
            if target-num in dic :
                return [dic[target-num], i]
            dic[num] = i
        return [-1, -1]


two_sum=lc001()
two_sum.n2([2, 7, 11, 15], 9)

##########  ##########  ##########  ##########  ##########

class lc013 : #E
    def n (self, s) :
        ans = 0
        values = {'M':1000, 'D':500, 'C':100, 'L':50, 'X':10, 'V':5, 'I':1}
        for i in range(0, len(s)-1) :
            c, c_nxt = s[i], s[i+1]
            if values[c] < values[c_nxt] :
                ans-=values[c]
            else :
                ans+=values[c]
        ans+=values[s[-1]]
        return ans


##########  ##########  ##########  ##########  ##########

class lc021 : #E
    def mergeTwoLists(self, lst1, lst2) :
        dummy=tail=ListNode(-1)
        while lst1 and lst2 :
            if lst1.val < lst2.val :
                tail.nxt=lst1
                lst1=lst1.nxt
            else :
                tail.nxt=lst2
                lst2=lst2.nxt
            tail=tail.nxt
        if lst1 :
            tail.nxt=lst1
        if lst2 :
            tail.nxt=lst2
        return dummy.nxt

##########  ##########  ##########  ##########  ##########

class lc026 : #E
    def removeDuplicates(self, nums: List[int]) -> int:
        n = len(nums)
        if n<= 1 : return n
        
        idx, pre = 1, nums[0]

        for i in range(1, len(nums)) :
            if pre == nums[i] : n-=1
            else :
                nums[idx] = nums[i]
                idx+=1
            pre = nums[i]
        
        return n

##########  ##########  ##########  ##########  ##########

class lc027 : #E, Remove Element
    def removeElement(self, nums: List[int], val: int) -> int:
        idx = 0
        for i in range(len(nums)) :
            if nums[i] != val :
                nums[idx] = nums[i]
                idx+=1
        return idx

##########  ##########  ##########  ##########  ##########

class lc066 : #E, Plus One
    def plusOne(self, digits: List[int]) -> List[int]:
        carry = 1
        for i in reversed(range(len(digits))) :
            d = (digits[i]+carry)%10
            carry=1 if d<digits[i] else 0
            digits[i]=d
        if carry :
            return [1]+digits
        return digits

#########  ##########  ##########  ##########  ##########

class lc088 : #E, Merge Sorted Array
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        """
        idx = m+n-1
        m, n = m-1, n-1
        
        while idx>=0 and m>=0 and n>=0 :
            if nums1[m] > nums2[n] :
                nums1[idx]=nums1[m]
                m-=1
            else :
                nums1[idx]=nums2[n]
                n-=1
            idx-=1
            
        while n>=0 :
            nums1[idx]=nums2[n]
            idx, n = idx-1, n-1

##########  ##########  ##########  ##########  ##########

class lc141 : #E
    def n(self, head) :
        fast=slow=head
        while fast and fast.nxt :
            fast = fast.nxt.nxt
            slow = slow.nxt
            if fast == slow :
                return True
        
        return False

##########  ##########  ##########  ##########  ##########

class lc561 : #E
    def n(self, lst) :
        total=0
        for i,x in enumerate(sorted(lst)) :
            if not i%2 :
                total += x
        return total

        return sum([c for i,c in enumerate(sorted(lst)) if not i%2])

##########  ##########  ##########  ##########  ##########

class lc566 : #E
    def rc(self, nums, r, c) :
        m=len(nums)
        n=len(num[0])
        if not m*n : return nums
        if r*c != m*n : return nums
        
        res = [[0]*c for _ in range(r)]
        for i in range(r*c) :
            res[i//c][i%c] = nums[i//n][i%n]
        
        return res


##########  ##########  ##########  ##########  ##########

class lc747 : #E
    def n(self, lst) :
        max1 = max2 = -1
        for i,x in enumerate(lst) :
            if x>max1 : 
                max1, max2 = x, max1
            elif x>max2 : 
                max2=x
        if max1 > 2*max2 : return True
        else : return False
        

########## Medium ##########  ##########  ##########  ##########

class lc002 : #M
    def n (self, lst1, lst2) :
        dummy=pivot=ListNode(-1)
        carry=0
        while lst1 or lst2 or carry :
            val = (lst1 and lst1.val or 0) + (lst2 and lst2.val or 0) + carry
            pivot.nxt = ListNode(val%10)
            carry = val//10
            lst1 = lst1 and lst1.nxt
            lst2 = lst2 and lst2.nxt
            pivot = pivot.nxt
        return dummy.nxt

lst_add=lc002()
list_to_num(lst_add.n(num_to_list(342),num_to_list(465)))

##########  ##########  ##########  ##########  ##########

class lc011 : #M
    def n (self, height) :
        l, r, ans = 0, len(height)-1, 0
        while l<r :
            ans = max(ans, (r-l)*min(height[l], height[r]))
            if height[l]<=height[r] : l+=1
            else : r-=1
        return ans

##########  ##########  ##########  ##########  ##########

class lc012 : #M
    def c (self, num) :
        ans = ""
        values  = {'M':1000, 'D':500, 'C':100, 'L':50, 'X':10, 'V':5, 'I':1}
        literals= ['M', 'D', 'C', 'L', 'X', 'V', 'I']
        for idx in [0,2,4] :
            quo = num//values[literals[idx]]
            rem = (num%values[literals[idx]])//values[literals[idx+2]]

            ans += quo*literals[idx]
            if rem==9 : ans += literals[idx + 2] + literals[idx]
            elif rem==4 : ans += literals[idx + 2] + literals[idx + 1]
            elif rem>=5 : ans += literals[idx + 1] + (rem - 5) * literals[idx + 2]
            else : ans += rem * literals[idx + 2]
            
            num %= values[literals[idx + 2]]
        return ans

##########  ##########  ##########  ##########  ##########

class lc015 : #M, 3sum
    def nlogn (self, nums) :
        ans = []
        nums.sort()
        for i in range(0,len(nums)) :
            if i>0 and nums[i]==nums[i-1] : # only unique triplets allowed
                continue
            two_sum, l_idx, r_idx = 0-nums[i], i+1, len(nums)-1
            while l_idx<r_idx :
                if nums[l_idx]+nums[r_idx] > two_sum :
                    r_idx-=1
                elif nums[l_idx]+nums[r_idx] < two_sum :
                    l_idx+=1
                else :
                    ans.append((nums[i], nums[l_idx], nums[r_idx]))
                    l_idx, r_idx = l_idx+1, r_idx-1
                    while l_idx<r_idx and nums[l_idx]==nums[l_idx-1] : l_idx+=1
                    while l_idx<r_idx and nums[r_idx]==nums[r_idx+1] : r_idx-=1
        return ans

##########  ##########  ##########  ##########  ##########

class lc016 : #M, 3sum-closest
    def nlogn(self, nums, target):
        ans, diff = 0, float("inf")
        nums.sort()
        for i in range(0,len(nums)) :
            l_idx, r_idx = i+1, len(nums)-1
            while l_idx<r_idx :
                three_sum=nums[i]+nums[l_idx]+nums[r_idx]
                dist = abs(three_sum-target)
                if dist < diff : 
                  diff, ans = dist, three_sum
                if three_sum > target :
                    r_idx-=1
                else :
                    l_idx+=1
        return ans

##########  ##########  ##########  ##########  ##########

class lc018 : #M, 4sum
    def fourSum(self, nums, target) :
        ans = []
        nums.sort()
        for i in range(0,len(nums)) :
            if i>0 and nums[i]==nums[i-1] :
                continue
            for j in range(i+1,len(nums)) :
                if j>0 and nums[j]==nums[j-1] :
                    continue
                
                two_sum, l_idx, r_idx = target-nums[i]-nums[j], j+1, len(nums)-1
                while l_idx<r_idx :
                    if nums[l_idx]+nums[r_idx]<two_sum :
                        l_idx+=1
                    elif nums[l_idx]+nums[r_idx]>two_sum :
                        r_idx-=1
                    else :
                        ans.append((nums[i],nums[j],nums[l_idx],nums[r_idx]))
                        l_idx,  r_idx = l_idx+1, r_idx-1
                        while l_idx<r_idx and nums[l_idx]==nums[l_idx-1] : l_idx+=1
                        while l_idx<r_idx and nums[r_idx]==nums[r_idx+1] : r_idx-=1
        return ans

##########  ##########  ##########  ##########  ##########

class lc019 : #M,
    def removeNthFromEnd(self, head, n) :
        slow=fast=dummy=ListNode(-1)
        dummy.nxt=head
        
        while n and fast :
            fast=fast.nxt
            n-=1
        
        if not fast : return dummy.nxt
        
        while slow.nxt and fast.nxt :
            slow=slow.nxt
            fast=fast.nxt
        
        slow.nxt=slow.nxt.nxt
        return dummy.nxt

##########  ##########  ##########  ##########  ##########

class lc024 : #M, lc025 #H
    def swapPairs(self, head: ListNode) -> ListNode:
        
        # https://www.geeksforgeeks.org/reverse-a-linked-list/
        # https://www.geeksforgeeks.org/reverse-a-list-in-groups-of-given-size/
        def reverseList(head, k):
            pre, cur, nxt, cnt = None, head, None, 0

            # ======================= #
            # check if has enough k node in this group
            # if not enough, don't do reverse
            while cur :
                cur = cur.nxt
                cnt += 1
            
            if cnt < k : return head
            else : cur, cnt = head, 0
            # ======================= #

            while cur and cnt<k :
                nxt = cur.nxt
                cur.nxt = pre
                pre, cur = cur, nxt
                cnt+=1

            if nxt:
                head.nxt = reverseList(nxt, k)

            return pre

        return reverseList(head, 2)

##########  ##########  ##########  ##########  ##########

class lc031 : #M,
    def nextPermutation(self, nums: List[int]) -> None:

##########  ##########  ##########  ##########  ##########

class lc036 : #M,
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        _row = [[0]*9 for _ in range(0,9)]
        _col = [[0]*9 for _ in range(0,9)]
        _box = [[0]*9 for _ in range(0,9)]
        
        for row_idx in range(0,9):
            for col_idx in range(0,9):
                if board[row_idx][col_idx] == '.' : 
                    continue
                val = int(board[row_idx][col_idx])-1
                box_idx = (row_idx//3)*3+(col_idx//3)
                if _row[row_idx][val] or _col[col_idx][val] or _box[box_idx][val] : 
                    return False
                _row[row_idx][val] = _col[col_idx][val] = _box[box_idx][val] = True
        return True

########## Hard ##########  ##########  ##########  ##########

class lc004 : #H
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:

##########  ##########  ##########  ##########  ##########

class lc023 : #H
    def mergeKLists_heap(self, lists):
        import heapq
        heap = []
        dummy = tail = ListNode(-1)
        [heapq.heappush(heap, (n.val, i)) for i, n in enumerate(lists) if n]

        while heap:
            val, lists_idx = heapq.heappop(heap)
            tail.nxt = node = lists[lists_idx]
            tail = tail.nxt
            if node.nxt:
                lists[lists_idx] = node = node.nxt
                heapq.heappush(heap, (node.val, lists_idx))
        return dummy.nxt
    
    def mergeKLists_merge(self, lists):
        
        def mergeTwoLists(lst1, lst2) :
            dummy=tail=ListNode(-1)
            while lst1 and lst2 :
                if lst1.val < lst2.val :
                    tail.nxt=lst1
                    lst1=lst1.nxt
                else :
                    tail.nxt=lst2
                    lst2=lst2.nxt
                tail=tail.nxt
            if lst1 :
                tail.nxt=lst1
            if lst2 :
                tail.nxt=lst2
            return dummy.nxt
      
      dummy = lists[0]
      for i in range(1, len(lists)) :
            dummy = mergeTwoLists(dummy, lists[i])
      return dummy

##########  ##########  ##########  ##########  ##########

class lc037 : #H
    def solveSudoku(self, board: List[List[str]]) -> None:
        _row = [[0]*9 for _ in range(9)]
        _col = [[0]*9 for _ in range(9)]
        _box = [[0]*9 for _ in range(9)]

        for row_idx in range(len(board)) :
            for col_idx in range(len(board)) :
                if board[row_idx][col_idx] != '.' :
                    box_idx = (row_idx//3)*3 + col_idx//3
                    val = int(board[row_idx][col_idx]) - 1
                    _row[row_idx][val]=_col[col_idx][val]=_box[box_idx][val]=True

        self.board, self._row, self._col, self._box = board, _row, _col, _box

        def _backtrack(row_idx, col_idx) :
            board, _row, _col, _box = self.board, self._row, self._col, self._box

            if board[row_idx][col_idx] != '.' :
                if row_idx==col_idx==len(board)-1 : return True
                if col_idx + 1 < len(board):
                    if _backtrack(row_idx, col_idx+1) : return True
                elif row_idx+1 < len(board) :
                    if _backtrack(row_idx+1, 0) : return True
            else :
                box_idx = (row_idx // 3) * 3 + col_idx // 3
                for val in range(0,9) :
                    if _row[row_idx][val]==True or _col[col_idx][val]==True or _box[box_idx][val]==True :
                        continue

                    # make move
                    _row[row_idx][val] = _col[col_idx][val] = _box[box_idx][val] = True
                    board[row_idx][col_idx] = str(val+1)

                    # backtrack
                    if row_idx==col_idx==len(board)-1 : return True
                    if col_idx+1 < len(board) :
                        if _backtrack(row_idx, col_idx+1) : return True
                    elif row_idx+1 < len(board) :
                        if _backtrack(row_idx+1, 0) : return True

                    # unmake move
                    _row[row_idx][val] = _col[col_idx][val] = _box[box_idx][val] = False
                    board[row_idx][col_idx] = '.'

            return False

        _backtrack(0,0)

##########  ##########  ##########  ##########  ##########

class lc041 : #H
    def firstMissingPositive(self, nums: List[int]) -> int:
        i = 0
        while i<len(nums) :
            if 0<nums[i]<=len(nums) and nums[i]!=nums[nums[i]-1] :
                # !!! INCORRECT !!! 
                # nums[i], nums[nums[i] - 1] = nums[nums[i] - 1], nums[i]
                # nums[i] will be modified first therefore affect the value of nums[nums[i] - 1]
                nums[nums[i] - 1], nums[i] = nums[i], nums[nums[i] - 1]
            else :
                i+=1
        
        for i in range(len(nums)) :
            if nums[i] != i+1 :
                return i+1
        
        return len(nums)+1


##########  ##########  ##########  ##########  ##########

class lc042 : #H, Trapping Rain Water
    def trap(self, height: List[int]) -> int:
        l_idx, r_idx, ans = 0, len(height)-1, 0
        l_max, r_max = float('-inf'), float('-inf')

        while l_idx<=r_idx :
            # 從左右兩邊開始比我當前高度低的才能積水
            if l_max<=r_max :
                ans += max(0,l_max-height[l_idx])
                l_max = max(l_max, height[l_idx])
                l_idx+=1
            else :
                ans += max(0,r_max-height[r_idx])
                r_max = max(r_max, height[r_idx])
                r_idx-=1
        return ans


##########  ##########  ##########  ##########  ##########

File: 001_array_linked_list.cpp
#include "leetcode.h"

//Maximum Gap
//Given an unsorted array, find the maximum difference between the successive elements in its sorted form.
class lc164 { //Hard
public:
    int MaximumGap(vector<int>& nums) {
        //Bucket sort. Use n buckets to store all the numbers. For each bucket, only track the min / max value.
        int n=nums.size();
        if(n<2) return 0;

        vector<int> min_bin(n,INT_MAX);
        vector<int> max_bin(n,INT_MIN);
        auto mm=minmax_element(nums.begin(),nums.end());
        int bin_size=(*mm.second-*mm.first)/n+1;
        for(auto num:nums) {
            int idx=(num-*mm.first)/bin_size;
            min_bin[idx]=min(min_bin[idx], num);
            max_bin[idx]=max(max_bin[idx], num);
        }

        int max_gap=0;
        int pre_max=max_bin[0];
        for(int i=1;i<n;i++)
            if(min_bin[i]!=INT_MAX) {
                max_gap=max(max_gap, min_bin[i]-pre_max);
                pre_max=max_bin[i];
            }
        return max_gap;
    }
};

//Max Chunks To Make Sorted II
//the integers of the given array are not necessarily distinct,
//the input array could be up to length 2000, and the elements could be up to 10**8.
class lc768 { //Hard
public:
    int MaxChunksToMakeSortedII(vector<int>& arr) {
        //Reduce the problem to 769. by creating a mapping from number to index in the sorted array.
        vector<int> map(arr.size());
        iota(map.begin(),map.end(),0); // assign idx to the mapping, start from 0. iota:用顺序递增的值赋值指定范围内的元素
        sort(map.begin(),map.end(), // sort the idx according the value
             [&arr](int x,int y){return arr[x]==arr[y] ? x<y : arr[x]<arr[y];});
        // then the same as 769
        int max_val=0, chunk_num=0;
        for(int i=0;i<map.size();i++) {
            max_val=max(max_val, map[i]);
            if(max_val==i) chunk_num++;
        }
        return chunk_num;
    }
};

//Max Chunks To Make Sorted
//Given an array arr that is a permutation of [0, 1, ..., arr.length - 1], we split
//the array into some number of "chunks" (partitions), and individually sort each chunk.
//After concatenating them, the result equals the sorted array.
//What is the most number of chunks we could have made?
class lc769 { //Medium
public:
    int MaxChunksToMakeSortedI(vector<int>& arr) {
        int max_val=0, chunk_num=0;
        for(int i=0;i<arr.size();i++) {
            max_val=max(max_val, arr[i]);
            if(max_val==i) chunk_num++; //目前最大值所屬之index出現了，表示在這之前的都可以自成一個chunk
        }
        return chunk_num;
    }
};

//Toeplitz Matrix
//A matrix is Toeplitz if every diagonal from top-left to bottom-right has the same element.
//Given an M x N matrix, return True if and only if the matrix is Toeplitz.
class lc766 { //Easy
private:
    int m,n;
public:
    bool ToeplitzMatrix(vector<vector<int>>& matrix) {
        m=matrix.size(); if(!m) return false;
        n=matrix[0].size(); if(!n) return false;

        for(int y=1;y<m;y++)
            for(int x=1;x<n;x++)
                if(matrix[y][x]!=matrix[y-1][x-1]) return false;
        return true;
    }
};

//Two Sum
//Given an array of integers, return indices of the two numbers such that
//they add up to a specific target.
class lc001 { //Easy
public:
    vector<int> TwoSum(vector<int>& nums, int target) {
        //time(N), space(N)
        unordered_map<int, int> table;
        for(auto i=0; i<nums.size(); i++)
            table[nums[i]]=i;
        for(auto i=0; i<nums.size(); i++) {
            int diff=target-nums[i];
            if(table.count(diff)&&table[diff]!=i)
                return {i, table[diff]};
        }
        return {};
    }
};

//Largest Number At Least Twice of Others
//In a given integer array nums, there is always exactly one largest element.
//Find whether the largest element in the array is at least twice as much as
//every other number in the array.
class lc747 { //Easy
public:
    int LargestNumberAtLeastTwiceOfOthers(vector<int>& nums) {
        int max1=-1, max2;
        auto i=max_element(nums.begin(),nums.end());
        swap(max1, *i);
        max2=*max_element(nums.begin(),nums.end());
        *i=max1;
        return max1>=max2*2 ? distance(nums.begin(),i) : -1;
    }
};

//Reshape the Matrix
//You're given a matrix represented by a two-dimensional array,
//and two positive integers r and c representing the row number and column number
//of the wanted reshaped matrix, respectively.
//The reshaped matrix need to be filled with all the elements of the original matrix
//in the same row-traversing order as they were.
class lc566 { //Easy
public:
    vector<vector<int>> ReshapeTheMatrix(vector<vector<int>>& nums, int r, int c) {
        //time(MN), space(MN)
        vector<vector<int>> matrix(r, vector<int>(c,0));
        int m,n;

        if(nums.empty()) return nums;
        if((m=nums.size())*(n=nums[0].size()) != r*c) return nums;
        for(auto i=0; i<m*n; i++)
            matrix[i/c][i%c]=nums[i/n][i%n];
        return matrix;
    }
};

//Array Partition I
//Given an array of 2n integers, your task is to group these integers into
//n pairs of integer, say (a1, b1), (a2, b2), ..., (an, bn) which makes
//sum of min(ai, bi) for all i from 1 to n as large as possible.
class lc561 { //Easy
public:
    int ArrayPartitionI(vector<int>& nums) {
        //time(NlogN), space(1)
        int sum=0;
        //先sort過，每個搭配比它大一點的數，也就是sort後的下一個
        sort(nums.begin(),nums.end());
        for(auto i=0; i<nums.size(); i+=2)
            sum+=nums[i];
        return sum;
    }
};

// 90: transpose, vertically    swap(A[m,n], A[n,m]), swap(A[m,n], A[m, dim-1-n])
//180: vertically, horizontally swap(A[m,n], A[m, dim-1-n]), swap(A[m,n], A[dim-1-m, n])
//270: transpose, horizontally  swap(A[m,m], A[n,m]), swap(A[m,n], A[dim-1-m, n])
void _ARotateMatrix90degree(vector<vector<int>> matrix)
{
    int m,n,dim=matrix.size();
    // swap(A[m,n], A[n,m]), transpose
    for(m=0; m<dim-1; m++)
        for(n=m+1; n<dim; n++)
            swap(matrix[m][n],matrix[n][m]);
    // swap(A[m,n], A[m, dim-n]), vertically
    for(m=0; m<dim; m++)
        for(n=0; n<dim/2; n++)
            swap(matrix[m][n],matrix[m][dim-1-n]);

    for(m=0; m<dim; m++) {
        for(n=0; n<dim; n++)
            cout<<matrix[m][n]<<' ';
        cout<<endl;
    }
}

//if an element is 0, set all its row/col to 0;
void _ASetRowColZero(vector<vector <int>> matrix)
{
    int i, j, row_p, col_p, n=matrix.size();
    bitset<4> row_z,col_z;

    row_p=col_p=1;
    for(i=0; i<n; i++) {
        for(j=0; j<n; j++) {
            row_p*=matrix[i][j];
            col_p*=matrix[j][i];
        }
        if(!row_p) row_z.set(i);
        if(!col_p) col_z.set(i);
        row_p=col_p=1;
    }

    for(i=0; i<n; i++) {
        for(j=0; j<n; j++) {
            if(row_z.test(i)||col_z.test(j))
                matrix[i][j]=0;
            cout<<matrix[i][j]<<' ';
        }
        cout<<endl;
    }
}

//=============================================================================LinkList+

//Split Linked List in Parts
//Given a (singly) linked list with head node root,
//write a function to split the linked list into k consecutive linked list "parts".
class lc725 { //Medium
public:
    vector<ListNode*> SplitLinkedListInParts(ListNode* root, int k) {
        vector<ListNode*> res(k);
        ListNode *head, *prev;
        int len=0;

        head=root;
        while(head) {len++;head=head->next;}

        int q=len/k; int r=len%k;
        head=root;prev=nullptr;
        for(int i=0;i<k;i++,r--) {
            res[i]=head;
            for(int j=0;j<q+(r>0);j++) {
                prev=head;
                head=head->next;
            }
            if(prev) prev->next=nullptr;
        }
        return res;
    }
};

//Merge Two Sorted Lists
//Merge two sorted linked lists and return it as a new list.
//The new list should be made by splicing together the nodes of the first two lists.
class lc021 { //Easy
public:
    ListNode* MergeTwoSortedLists(ListNode* l1, ListNode* l2) {
        //time(N), iterative
        ListNode head(0);
        ListNode *tail=&head;
        while(l1&&l2) {
            if(l1->val<l2->val) { tail->next=l1; l1=l1->next; }
            else { tail->next=l2; l2=l2->next; }
            tail=tail->next;
        }
        if(l1) tail->next=l1; if(l2) tail->next=l2;
        return head.next;

        //==============================
        //time(N) recursion
        if(!l1||!l2) return l1?l1:l2;
        if(l1->val<l2->val) {l1->next=MergeTwoSortedLists(l1->next,l2); return l1;}
        else {l2->next=MergeTwoSortedLists(l1, l2->next); return l2;}
    }
};

//Linked List Cycle
//Given a linked list, determine if it has a cycle in it.
class lc141 { //Easy
public:
    bool LinkedListCycle(ListNode *head) {
        // time(N) space(N)
        unordered_set<ListNode*> visited;
        while(head) {
            if(visited.count(head)) return true;
            visited.insert(head);
            head=head->next;
        }
        return false;

        // fast/slow pointers, no need extra space
        ListNode *fast, *slow;
        fast=slow=head;
        while(fast) {
            if(!fast->next) return false;
            if((fast=fast->next->next)==(slow=slow->next)) return true;
        }
        return false;
    }
};

ListNode* _LFindLoopStart(ListNode* node)
{
//    if no duplicated key value, use hash table, like in BFS/DFS Graph traverse

//    k: length of non-looped part
//    n: length of looped part
//    x: length from loop start to meet point
//
//    Prove: k=n-x;
//    Known: 1st meet: 2x(steps of slower)=(steps of faster)
//    =>2(k+x)=(k+n+x)
//    =>k=n-x
    ListNode *slow, *fast;

    slow=fast=node;
    while(fast) { // find meeting point
        if(!fast->next) return nullptr; // no loop
        if((slow=slow->next)==(fast=fast->next->next)) break;
    }

    while(node!=fast) { // k=n-x
        node=node->next;
        fast=fast->next;
    }
    return node;
}

void _LRemoveDuplicate(ListNode *node)
{
    ListNode *pivot=node, *prev;

    // time(N^2), no temp buff
    while(pivot) {
        node=pivot->next;
        while(node) {
            if(pivot->val!=node->val) {
                prev=node;
                node=node->next;
            } else {
                prev->next=node->next;
                delete node;
                node=prev->next;
            }
        }
        pivot=pivot->next;
    }
    //==============================
    // time(N)
    bitset<100> table;
    while(node) {
        if(!table.test(node->val)) {
            table.set(node->val);
            prev=node;
            node=node->next;
        } else {
            prev->next=node->next;
            delete node;
            node=prev->next;
        }
    }
}
ListNode* _LFindLastNNode(ListNode* node, int n)
{
    ListNode *fr_node, *rr_node;

    if(!node||n<1) return nullptr;
    else fr_node=rr_node=node;
    for(auto i=0; i<n-1; i++) { // skip n-1 steps ahead
        if(!rr_node) return nullptr;
        rr_node=rr_node->next; // not found since list size < n
    }
    while(rr_node->next) {
        fr_node=fr_node->next;
        rr_node=rr_node->next;
    }
    return fr_node;
}

bool _LRemoveNode(ListNode* node)
{
    ListNode *nx_node;

    if(!node||!node->next) return false; //this algo can't handle the case of removing the last node

    nx_node=node->next;
    delete node;
    node->val=nx_node->val;
    node->next=nx_node->next;

    return true;
}

ListNode* _LAddTwoNumber(ListNode *x, ListNode *y, ListNode *sum)
{
    ListNode *result=sum;
    int carry=0;

    if(!x||!y||!sum) return nullptr;

    while(x||y) {
        if(x) {carry+=x->val; x=x->next;}
        if(y) {carry+=y->val; y=y->next;}
        sum->val=carry%10;carry/=10;
        sum=sum->next;
    }
    if(carry) sum->val=carry;
    return result;
}

//=============================================================================LinkList-
#if 0
int main()
{
    vector<vector<int>> m{{1,  2,  3,  4},
                          {5,  6,  7,  8},
                          {9,  0, 11, 12},
                          {13, 14, 15, 16}};
//    _ARotateMatrix90degree(m);
    _ASetRowColZero(m);

    ListNode *node, *root;
    node=root=new ListNode(1);
    while(node->next) node=node->next;
    node->next=new ListNode(2);
    node=root;
    while(node->next) node=node->next;
    node->next=new ListNode(3);
    node=root;
    while(node->next) node=node->next;
    node->next=new ListNode(6);
    node=root;
    while(node->next) node=node->next;
    node->next=new ListNode(2);

    node=root;
    while(node) {
        cout<<node->val<<' ';
        node=node->next;
    }
    node=root;
//    _LRemoveDuplicate(node);
//    node=_LFindLastNNode(node, 2);
//    _LRemoveNode(node3);
    node=_LAddTwoNumber(node,node,node);
//    node=root;
    cout<<endl;
    while(node) {
        cout<<node->val<<' ';
        node=node->next;
    }
    cout<<endl<<"Hello 02";
    return 0;
}
#endif
